tools/setup_services.py

```diff
@@ -1,15 +1,396 @@
+        plugin_path (str): Directory where the command runs
+        plugin_name (str): Used in error logs
+
+    Returns:
+        stdout_lines (list of str): Output lines from stdout
+        stderr_str (str): Full stderr output
+
+    Raises:
+        subprocess.CalledProcessError: If the command fails
+    """
+    try:
+        result = subprocess.run(
+            ["git"] + args,
+            cwd=plugin_path,
+            capture_output=True,
+            text=True,
+            check=True,
+        )
+
+        return result.stdout.splitlines(), result.stderr
+
+    except subprocess.CalledProcessError as e:
+        msg = f"[ERROR] Git command failed: {' '.join(args)}"
+        if plugin_name:
+            msg += f" (plugin: {plugin_name})"
+        msg += f"\n{e.stderr}"
+        print(msg)
+        raise
+
+
+def download_plugin_services():
+    _, cfg = load_env()
+    plugins_path = cfg.get("services.plugins_path", "./plugins")
+    os.makedirs(plugins_path, exist_ok=True)
+
+    plugins = cfg.get("plugins", {})
+    plugin_services = []
+
+    log(f"Discovered {len(plugins)} plugins")
+
+    for plugin_name, plugin_info in plugins.items():
+        if "url" not in plugin_info:
+            log(f"Skipping plugin {plugin_name} because it has no URL")
+            continue
+
+        if process_plugin(plugin_name, plugin_info, plugins_path):
+            plugin_services.append(plugin_name)
+
+    return plugin_services
+
+
+def process_plugin(plugin_name, plugin_info, plugins_path):
+    """Process a single plugin - clone or update as needed."""
+    plugin_path = pjoin(plugins_path, plugin_name)
+
+    if os.path.exists(plugin_path):
+        success = update_existing_plugin(plugin_name, plugin_info, plugin_path)
+    else:
+        success = clone_new_plugin(plugin_name, plugin_info, plugin_path)
+
+    if success and validate_plugin_compatibility(plugin_name, plugin_path):
+        generate_supervisor_config(plugin_name, plugin_path)
+        return True
+
+    return False
+
+
+def update_existing_plugin(plugin_name, plugin_info, plugin_path):
+    """Update an existing plugin repository."""
+    if not is_valid_git_repo(plugin_path):
+        return False
+
+    if has_modified_files(plugin_path, plugin_name):
+        return False
+
+    version_tag = plugin_info.get("version")
+    sha = plugin_info.get("sha")
+    branch = plugin_info.get("branch", "main")
+
+    if version_tag:
+        return update_to_version(plugin_name, plugin_path, version_tag)
+    elif sha:
+        return update_to_sha(plugin_name, plugin_path, sha)
+    else:
+        return update_to_branch(plugin_name, plugin_path, branch)
+
+
+def clone_new_plugin(plugin_name, plugin_info, plugin_path):
+    """Clone a new plugin repository."""
+    branch = plugin_info.get("branch", "main")
+    version_tag = plugin_info.get("version")
+    sha = plugin_info.get("sha")
+
+    log(f"Cloning plugin {plugin_name}")
```

Comment:
Instead of doing a deep clone, can we do a shallow clone of exactly the revision specified? If so, should be able to get rid of the three `update_to_` functions.